### ISSUE TITLE
Camera preview takes time to open in ios after capacitor 3 upgrade

### ISSUE THREAD

**Issue by yksolanki9** [capacitor-community/camera-preview#218]
**Describe the bug**
I have been using this plugin for a while, but recently we upgraded from capacitor 2 to 3 and upgraded the plugin to v3 and while things work as usual in android, we have been experiencing some delay while opening the camera preview in ios. Specifically, on clicking on the camera cta, we see a black screen for a while and then the camera preview opens after about 2-4 seconds. I've tested this on iphone 6(ios 12) and iphone XR(ios 15) and facing this issue in both the devices.
And like I said things work fine on android devices

**Comment by yksolanki9** [comment]
For capacitor 2, there was no noticable lag in opening the camera preview. But since upgrading to capacitor 3 and plugin v3, the delay has increased considerably

**Comment by xavierchamiotbasset** [comment]
Hi, any news on this topic ? I experience the same issue as described : the camera preview takes more than 2 seconds to open on IOS only. Tested on Iphone SE (ios15) and Iphone 15 (ios15).

**Comment by Sparkfield** [comment]
Since I have the same problem, I did some debugging on this issue.
It seems to be related to AVCaptureSession in the file CameraController, especially when calling startRunning() and doing some stuff in func configureDataOutput().

Looking into the Apple documentation, I found information about this issue:
[https://developer.apple.com/documentation/avfoundation/avcapturesession](https://developer.apple.com/documentation/avfoundation/avcapturesession)

This page says:

> Important
> 
> The startRunning() method is a blocking call which can take some time, therefore start the session on a serial dispatch queue so that you don’t block the main queue (which keeps the UI responsive). See [AVCam: Building a Camera App](https://developer.apple.com/documentation/avfoundation/capture_setup/avcam_building_a_camera_app) for an implementation example.

I don't know anything about Swift, so I don't know how to make the appropriate changes in order to make this work with a serial dispatch queue. One thing I do know is that the current code does not use the serial dispatch queue.

Hopefully someone with Swift knowledge can look further into this issue.

**Comment by Sparkfield** [comment]
I dived further into the code and noticed that some dispatching is already taking place, but all logic is handled in one dispatch queue (DispatchQueue(label: "prepare").async).

I added some code to measure the duration of specific functions. These are the results:

Time elapsed for configureCaptureDevices: 0.01330411434173584 s.
Time elapsed for configureDeviceInputs: 0.003573894500732422 s.
Time elapsed for configurePhotoOutput: 0.5769590139389038 s.
Time elapsed for configureDataOutput: 0.3801349401473999 s.
Time elapsed for updateVideoOrientation: 0.45588791370391846 s.

When adding all the timestamps, the total time is about 1.5 seconds to execute.
Is it possible to make this code run concurrently to bring down the time, maybe by adding a few extra dispatchQueues?

Anyone here who can investigate this further?

**Comment by THSoftware** [comment]
Im having the same problem on iOS. Preview takes around 2 seconds to load every time, it's causing quite an issue. Can anyone please help?

**Comment by ygyg70** [comment]
I am having the same issue :(

**Comment by michaelwolz** [comment]
@ygyg70 in case you don't need anything more than just using the plugin for taking photos you can try out my fork: https://www.npmjs.com/package/@michaelwolz/camera-preview-lite 

(I removed a few functionalities which I didn't need though)

**Comment by ygyg70** [comment]
Thanks, I did look at it however I am using to take repeated snapshots and the experience is better with the original plugin.
https://capawesome.io/plugins/mlkit/barcode-scanning/ is doing something similar and takes a split second to start, so I am assuming there is a better way. If my project will pick up I may have time to invest here.

**Comment by michaelwolz** [comment]
Yes that's right. I also took a look into the barcode scanning plugin when I tried to resolve this issue because it actually opens immediately. What actually takes so much time, is the configuration of the camera outputs which is required for taking photos. 

See: 
 - https://github.com/capacitor-community/camera-preview/blob/d048417c97589cfbed412016fe8e07cf8276cfbc/ios/Plugin/CameraController.swift#L139-L140 
 - or in my case: https://github.com/michaelwolz/camera-preview-lite/blob/4f43d1186a5efd89cb75427fe164efe7c498815c/ios/Plugin/CameraController.swift#L82-L85). 

If you remove this code the camera will start up almost immediately (but you won't be able to take photos then).

As far as I remember, the barcode scanning plugin doesn't need this because it directly acts on the video stream. If you use lower quality on the photo output it will also be a little faster. I also recognized that the issue is getting worse the better the camera of the phone is. So an iPhone 11 opens up the camera faster than iPhone 14 Pro, for example. I am not too deep into swift and the AVFoundation framework. I am pretty sure there is a better solution utilizing a queue to outsource initialization of the photo output somehow. So in case anybody is having an idea I'd also be happy to try it out :).
